### ISSUE TITLE
Nix caches (locally) 404 responses received from (remote) binary caches

### ISSUE THREAD

**Issue by ryantrinkle** [NixOS/nix#781]
I maintain a binary cache for my project, ryantrinkle/try-reflex, and sometimes a user will try to build that project before I have cached all the binaries for it.  If I then add more packages to the binary cache, that user won't be able to get those files until they `rm /nix/var/nix/binary-cache-v3.sqlite*`.

**Comment by witkamp** [comment]
I think this is related to issue https://github.com/NixOS/nix/issues/396

**Comment by vcunat** [comment]
I thought it was on purpose that negative answers are cached as well.

**Comment by ryantrinkle** [comment]
@vcunat Hmm.  Well, if it's intentional, I'm perfectly happy to close the ticket, but I must say that it's quite surprising behavior, at least to me.

**Comment by copumpkin** [comment]
It seems like we might want negative answers to expire fairly quickly, or for the server to be able to specify some sort of expiry policy for its binary cache so clients know how it gets updated.

**Comment by ryantrinkle** [comment]
Yes, having a reasonably quick expiration time would make sense to me.  I'm not sure how feasible it is, but I suppose the ideal would be to respect whatever caching headers the server supplies in the response.

**Comment by cleverca22** [comment]
related, i have noticed that nix disables the journal when accessing nix/var/nix/binary-cache, which can lead to a corrupt cache database, its fine since you can just delete it and it auto-rebuilds

but in one case, nix just hard-failed all operations, and didn't clean the mess up itself, so i had to debug the code, and then delete that file

**Comment by edolstra** [comment]
Currently the TTL for negative queries is hardcoded in `scripts/download-from-binary-cache.pl.in` at 1 hour. Perhaps this should be overridable via the `nix-cache-info` file of the binary cache.

**Comment by domenkozar** [comment]
Relevant: https://github.com/NixOS/nix/commit/74dd603495273fe8b81d8635676861792cf420e8
